### PR TITLE
DistinctOperation now returns a batch Cursor

### PR DESCRIPTION
**Pull request by rozza**
JAVA-1624

### CODE REVIEW

File: bson/src/main/org/bson/codecs/BsonArrayCodec.java [hunk 1]
```diff
@@ -21,26 +21,41 @@
...
 public class BsonArrayCodec implements Codec<BsonArray> {
-    private final CodecRegistry registry;
+    private static final CodecRegistry DEFAULT_REGISTRY = new RootCodecRegistry(Arrays.<CodecProvider>asList(new BsonValueCodecProvider()));
+
+    private final CodecRegistry codecRegistry;
+
+    /**
+     * Creates a new instance with a default {@link org.bson.codecs.configuration.RootCodecRegistry}
+     */
+    public BsonArrayCodec() {
```
Comment (jyemin) [inline bson/src/main/org/bson/codecs/BsonArrayCodec.java]: This constructor appears to be unused.  Why did you go to the trouble of creating it?

Comment (rozza) [inline bson/src/main/org/bson/codecs/BsonArrayCodec.java]: Follows `BsonDocumentCodec` strictly speaking I'm not sure we need to pass an alternative codecRegistry in.

File: driver-core/src/main/com/mongodb/operation/BsonArrayWrapper.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+
+import static org.bson.assertions.Assertions.notNull;
+
+
+class BsonArrayWrapper<T> extends BsonArray {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/BsonArrayWrapper.java]: Most of this class appears to be unused and untested.  In particular, getUnwrapped is never called.

File: driver-core/src/main/com/mongodb/operation/BsonArrayWriter.java [hunk 3]
```diff
@@ -0,0 +1,241 @@
...
+import org.bson.BsonUndefined;
+import org.bson.BsonValue;
+import org.bson.BsonWriterSettings;
+import org.bson.types.ObjectId;
+
+import static org.bson.BsonContextType.DOCUMENT;
+import static org.bson.BsonContextType.SCOPE_DOCUMENT;
+
+
+class BsonArrayWriter extends AbstractBsonWriter {
```
Comment (jyemin) [inline driver-core/src/main/com/mongodb/operation/BsonArrayWriter.java]: Called by unwrapped above, but never tested.

**Comment by jyemin** [comment]
Looks good.  As for the un-exercised code, since it's all in non-public classes, another option (besides just writing the tests) is to just remove it and replace with throw new UnsupportedOperationException().  Otherwise,

**Comment by jyemin** [comment]
Proposing a simplification of BsonValueCodec, as I found its dependency on BsonDocumentCodec confusing.

**Comment by rozza** [comment]
Looks better thanks :+1: 

I've updated and `BsonArrayWrapper` is not much of a wrapper but we can always implement the missing features should the need ever arise.  I changed the `BsonWrapperHelper.toList` method signature to try and make it more palatable.

**Comment by jyemin** [comment]
lgtm

**Comment by rozza** [comment]
rebased an merged

**Pull request closed without merging**